app/services/mentions/create_all.rb

```diff
@@ -48,29 +48,38 @@ def extract_usernames_from_mentions_in_text
     end
 
     def reject_notifiable_author(users)
-      users.reject { |user| authored_by?(user, @notifiable) }
+      users.reject { |user| authored_by?(user, notifiable) }
     end
 
     def authored_by?(user, notifiable)
       user.id == notifiable.user_id
     end
 
     def delete_mentions_removed_from_notifiable_text(users)
-      mentions = @notifiable.mentions.where.not(user_id: users).destroy_all
+      mentions = notifiable.mentions.where.not(user_id: users).destroy_all
       Notification.remove_all(notifiable_ids: mentions.map(&:id), notifiable_type: "Mention") if mentions.present?
     end
 
     def user_has_comment_notifications?(user)
-      user.notifications.exists?(notifiable_id: @notifiable.id)
+      user.notifications.exists?(notifiable_id: notifiable.id, notifiable_type: "Comment")
     end
 
     def create_mention_for(user)
-      return if user_has_comment_notifications?(user)
+      # Do not create additional notifications for being mentioned in a comment.
+      if notifiable.is_a?(Comment) && user_has_comment_notifications?(user)
+        return
+      end
+
+      # The mentionable_type is the model that created the mention, the user is the user to be mentioned.
+      mention = Mention.create(user_id: user.id, mentionable_id: notifiable.id,
+                               mentionable_type: notifiable.class.name)
+
+      # If notifiable is an Article, we need to create the notification for the mention immediately so
+      # that the notification exists in the database before we attempt to create other Article-related notifications.
+      # However, if notifiable is a Comment, we can create the notification for the mention in the background.
+      Notification.send_mention_notification_without_delay(mention) if notifiable.is_a?(Article)
```

Comment:
As-is, this will always perform at least one unnecessary check.

1. If `notifiable` is an `Article` the next line will still check if it is a comment.
2. If `notifiable` is a `Comment` it will have performed the check for `Article` nonetheless.

There's not much you can do about 2 but we can do something about 1:

```ruby
case notifiable
when Article
  Notification.send_mention_notification_without_delay(mention)
when Comment
  Notification.send_mention_notification(mention)
end
```

The reason I chose `case` over `if/elsif` is that AR classes have a proper `===` implementation [1] (which is what's used for case-equality) so you can get nice conditionals but they perform exactly the same check:

```ruby
[9] forem(main)> puts Comment.method(:===).source
      def ===(object) # :nodoc:
        object.is_a?(self)
      end
```

BTW, this is not intended as a micro-optimization but to enhance readability. The second `if` really threw me for a second because my brain was like "this can never be the case if the previous line was executed".

[1] It's slightly different from the one defined in `Module` which is implemented in C as `return rb_obj_is_kind_of(arg, mod);`,